ext/ddtrace_profiling_native_extension/linux_tid_from_pthread.c

```diff
@@ -0,0 +1,221 @@
+// 3. We follow that pointer and all goes well, we expect that at some offset from the memory `pthread_t` was pointing
+//    at, we'll find the reference thread's tid.
+// 4. Since the offset for every thread is the same (the layout of `struct pthread` doesn't change), we can now use this
+//    offset to get the tid any other thread we want, given its `pthread_t`.
+//
+// And after step 4, if we were successful, we can call `linux_tid_from` with the offset to get the tid for any thread.
+//
+// #### How to make this safe
+//
+// You may be asking yourself, isn't what is described above incredibly unsafe and crashy in C? The answer is yes, if
+// not done carefully. Specifically, it's crashy if we just cast `pthread_t` to a `void *`, dereference it, and hope for
+// the best. But that's not what we're doing.
+//
+// What we do instead is use the very powerful `process_vm_readv` Linux system API that allows you to read memory
+// from a process in a safe manner -- e.g., it either succeeds or it returns a failure status code, rather than
+// triggering a segfault. (This API is actually able to read memory from any process, and is usually used to read memory
+// from other processes, but can actually be used to read memory from the current process in this safe way).
+//
+// Thus, we can safely read the opaque block of memory that we suspect exists at that memory location, and search it for
+// the info we want.
+//
+// Note: As documented in the man page, `process_vm_readv` needs special permissions and may not work on every Linux
+// setup. This is handled below as any other error, and just makes this approach not usable in those machines.
+//
+// #### How to make this accurate
+//
+// After reading memory safely, you may wonder -- what if the tid is really low, like 1 or 2 (e.g. in a docker container)
+// and there's other things inside `struct pthread` that may be set to 1 or 2? Won't that cause us to find the wrong
+// offset?
+//
+// To solve this issue, inside `setup_linux_tid_from_pthread_offset` we actually don't use only the current thread as
+// a reference thread -- we actually use REFERENCE_THREADS_COUNT (3, as of this writing) threads as a reference.
+//
+// The thinking is -- the odds are really low that multiple different threads have, at the same offset, a value that is
+// exactly same as their tid, but not their tid.
+//
+// That's why we document in the `linux_tid_from_pthread.h` that it's rare-but-not-theoretically-impossible that the
+// result we get is wrong.
+//
+// ---
+
+#ifdef __linux__
+
+  #define _GNU_SOURCE
+  #include <pthread.h>
+  #include <stdbool.h>
+  #include <stdio.h>
+  #include <string.h>
+  #include <sys/syscall.h>
+  #include <sys/uio.h>
+  #include <unistd.h>
+  #include "linux_tid_from_pthread.h"
+
+  // This value seems enough for both glibc and musl (and it's safe to read with `process_vm_readv`)
+  // Must be a multiple of sizeof(pid_t)
+  #define STRUCT_PTHREAD_READ_SIZE 2048
```

Comment:
Can we document why `2048`?
From the current comment, I'm not sure how large glibc's and musl's versions are at the time of this writing (for an example architecture).
Also, links to glibc and musl data structures would be helpful here.

My main thing here is: I'm not sure what 2048 is supposed to be. Is it close to glibc's struct size? is it double, just to be safe? Is it like 10x more, to be even more safe? is it half the size, to avoid reading garbage?

Reply:
Extremely fair! In hindsight I agree I was handwaving too much here. I've added this as a comment in b31620734c13bbdf59f7b9710f9b23537bd0c2f8 :

````
In particular, I've picked 2048 after examining the layouts of the structures using gdb. This value is a bit more
than double the needed for glibc, and is way higher than needed for alpine, so we have some leeway for the field
moving around in different libc versions and architectures.

* On Ubuntu 22.04's x86-64 glibc 2.35:
 ```
 gdb /lib/x86_64-linux-gnu/libc.so.6
 (gdb) print sizeof(struct pthread)
 $1 = 2496
 (gdb) ptype /o struct pthread
 /* offset      |    size */  type = struct pthread {
 /*    720      |       4 */    pid_t tid;
 ```
* On alpine 3.18.3's x86-64 musl 1.2.4:
  ```
 gdb /lib/libc.musl-x86_64.so.1
 (gdb) print sizeof(struct __pthread)
 $1 = 200
 (gdb) ptype /o struct __pthread
 /* offset      |    size */  type = struct __pthread {
 /*     48      |       4 */    int tid;
 ```

In the future, we may want to tune this down if we see that process_vm_readv fails because it can't read the whole
2048 bytes (e.g. on musl?), but it's a trade-off between overfitting and overreading :)
````

It's cool that gdb is able to print the structure and its layout :)